### PR TITLE
Bugfix for fish support: overly zealous arg matching

### LINKED ISSUE #18369: spack load with fish shell fails when package contains -h
### Steps to reproduce the issue

```console
~> spack load omega-h
==> This command works best with Spack's shell support
  
  To initialize spack's shell commands:
  
      # for bash and zsh
      . /home/wortiz/spack/share/spack/setup-env.sh
  
      # for csh and tcsh
      setenv SPACK_ROOT /home/wortiz/spack
      source /home/wortiz/spack/share/spack/setup-env.csh
  
  Or, if you want to use `spack load` without initializing
  shell support, you can run one of these:
  
      eval `spack load --sh omega-h`   # for bash/sh
      eval `spack load --csh omega-h`  # for csh/tcsh

```

### Error Message

```console~> spack --debug --stacktrace load omega-h
lib/spack/spack/cmd/__init__.py:122 ==> [2020-08-28-15:31:53.406215] Imported load from built-in commands
lib/spack/spack/cmd/__init__.py:122 ==> [2020-08-28-15:31:53.407513] Imported load from built-in commands
lib/spack/spack/config.py:835 ==> [2020-08-28-15:31:53.410728] Reading config file /home/wortiz/spack/etc/spack/defaults/config.yaml
lib/spack/spack/database.py:361 ==> [2020-08-28-15:31:53.424942] DATABASE LOCK TIMEOUT: 3s
lib/spack/spack/database.py:365 ==> [2020-08-28-15:31:53.425115] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/config.py:835 ==> [2020-08-28-15:31:53.454480] Reading config file /home/wortiz/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:835 ==> [2020-08-28-15:31:53.456232] Reading config file /home/wortiz/.spack/repos.yaml
lib/spack/spack/cmd/load.py:76 ==> [2020-08-28-15:31:53.495821] This command works best with Spack's shell support
  
  To initialize spack's shell commands:
  
      # for bash and zsh
      . /home/wortiz/spack/share/spack/setup-env.sh
  
      # for csh and tcsh
      setenv SPACK_ROOT /home/wortiz/spack
      source /home/wortiz/spack/share/spack/setup-env.csh
  
  Or, if you want to use `spack load` without initializing
  shell support, you can run one of these:
  
      eval `spack load --sh omega-h`   # for bash/sh
      eval `spack load --csh omega-h`  # for csh/tcsh

```

### Additional information

Workaround:
```console
~> eval (spack load --fish omega-h)
```

Appears to be caused by ` function check_env_activate_flags` being too lax when checking for -h

Can be fixed with the following patch but I'm not sure if this is strict enough

```patch
diff --git a/share/spack/setup-env.fish b/share/spack/setup-env.fish
index dd2a17f74..9200d7136 100755
--- a/share/spack/setup-env.fish
+++ b/share/spack/setup-env.fish
@@ -258,7 +258,7 @@ function check_env_activate_flags -d "check spack env subcommand flags for -h, -
     # skip if called with blank input. Notes: [1] (cf. EOF)
     if test -n "$_a"
         # looks for a single `-h` (possibly surrounded by spaces)
-        if echo $_a | string match -r -q " *-h *"
+        if echo $_a | string match -r -q " +-h *"
             return 0
         end
 
```

I don't usually write fish programs so maybe someone more knowledgable knows of a better way to check args

@JBlaschke

### PR DESCRIPTION
**Pull request by JBlaschke**
@wortiz brought up an interesting issue with the current way we're detecting `-h` (and so on) in #18369 

This PR fixes that bug by matching `-h`, `--help`, `--sh`, `--csh`, and `--fish` _exactly_. The old way would identify `<anything>-h` and `-h<anything>`.

Fixes #18369

### CODE REVIEW

**Comment by adamjstewart** [comment]
Thanks! This was fixed on the sh/csh side in #15429, but that was after you had started your fish support and before it was merged.

File: share/spack/setup-env.fish
Comment (adamjstewart) [inline share/spack/setup-env.fish]:
```suggestion
    # Remove leading and trailing spaces -- but we need to insert a "guard" (x)
```

**Comment by JBlaschke** [comment]
@adamjstewart yep -- that's the price I pay for taking too long 😉

**Review by adamjstewart — APPROVED** [review]
I'll wait for one more person to review.

**Comment by adamjstewart** [comment]
> @adamjstewart yep -- that's the price I pay for taking too long 😉

Good things are worth the wait.

My new roommate is also a fish user. There's a good chance I'll be switching to fish in the near future, so I'll be able to file a lot more bug reports. But first I need to port all of my [dotfiles](https://github.com/adamjstewart/dotfiles) from bash to fish...

**Comment by JBlaschke** [comment]
> But first I need to port all of my [dotfiles](https://github.com/adamjstewart/dotfiles) from bash to fish...

This might not be a permanent solution (or at least not an elegant solution) but to get started quickly, you could give `bax` a try: https://github.com/jorgebucaran/bax.fish -- it is a bash execution wrapper that replays environment changes to your fish env.

**Pull request merged**
